Dev/metadata

Nic-Chr opened the pull request:
Added metadata functionality to `get_spd()`.

Nic-Chr committed:
Added spd metadata.

Nic-Chr committed:
Added metadata helper.

Nic-Chr committed:
New helpers and better messaging.

Nic-Chr committed:
Added helpful messages.

Nic-Chr committed:
Style code (GHA)

Moohan commented on R/get_spd.R:
```diff
@@ -53,8 +55,22 @@ get_spd <- function(version = "latest", col_select = NULL) {
+
+  cli::cat_line("\n--- Metadata ---\n", col = "blue")
+  cli::cat_print(metadata)
+  cli::cat_line("")
```
I think this is too much. I tried to limit the rows, but couldn't see how to do it with `cli::cat_print()`.

```
--- Metadata ---

# A tibble: 95 × 2
   variable             description                                                                        
   <chr>                <chr>                                                                              
 1 pc7                  The Royal Mail postcode in a 7 character format                                    
 2 pc8                  The Royal Mail postcode in a 8 character format                                    
 3 split_char           The split indicator added by NRS (blank if postcode is whole)                      
 4 pc_district          Postcode District                                                                  
 5 pc_sector            Postcode Sector                                                                    
 6 date_of_introduction Postcode introduction date                                                         
 7 date_of_deletion     Postcode deletion date                                                             
 8 postcode_type        Small or Large User postcode indicator                                             
 9 pc_su_link           Provides the Small User postcode that contains the grid reference of the Large Use…
10 split_su_link        The split indicator for small user link postcode added by NRS (blank if postcode i…
# ℹ 85 more rows
# ℹ Use `print(n = ...)` to see more rows
```

Maybe `dplyr::glimpse()` will give a shorter but still informative output that can be used?

Nic-Chr replied:
We can use `print(metadata, n = 5)` to print the first five rows?

Moohan commented on R/get_spd.R:
```diff
@@ -53,8 +55,22 @@ get_spd <- function(version = "latest", col_select = NULL) {
   }
 
-  return(read_file(
+  metadata <- readr::read_csv(fs::path(metadata_dir, "spd_metadata.csv")) %>%
```
Maybe have a check `fs::file_exists()` first, so if the metadata file we expect doesn't exist, we can either fail or skip (and warn?).

Additionally, we could possibly use `read_file` to read the metadata? That way, it's resilient to a future change of file type, and also any improvements we make to read_file would be included, for example, if we introduced caching etc.

Moohan reviewed (changes requested):
Works well for me. 

Most suggestions are around the messages and warnings in the console.

Moohan committed:
Merge branch 'main' into dev/metadata

Moohan committed:
Merge branch 'main' into dev/metadata

Nic-Chr committed:
Rename cols with readr

Nic-Chr committed:
Use base print

Nic-Chr committed:
Use read_file and added file check warning

Nic-Chr committed:
Style code (GHA)

bnowok committed:
Merge branch 'main' into dev/metadata

Moohan committed:
Merge branch 'main' into dev/metadata

Moohan committed:
Refactor metadata messaging into the metadata fucntions

Moohan committed:
Add an error if `metadata` is used on a non-tibble

Moohan committed:
Fix and improve tests to expect the messaging

Moohan committed:
Improve CSV lazy read and suppress progress messages

The progress message was being printed to the console as `0s` or `1s`, not required and potentially confusing.

CSV lazy reading is fine if the user is on UNIX (linux/mac), it's only not good when on Windows, so limit it to that. This should make reading CSV files (metadata + some of the older lookups) faster on Posit Workbench.

Moohan committed:
Style code (GHA)

Moohan committed:
Refactor to make loading the metadata be on demand

This offloads metadata loading to the first call of `metadata()`. When the lookup is loaded, it sets some attributes, such as where the metadata file is and whether it exists, but it's not loaded initially.

Overall, I think this is a better approach. The only downside is I couldn't figure out a good way to partially print the top few rows of the metadata, as that (obviously) involves reading it.

Moohan committed:
Merge pull request #51 from Public-Health-Scotland/dev/metadata-lazy-load

Refactor to make loading the metadata be on demand

Moohan committed:
Document package (GHA)

Moohan committed:
Add error handling for missing metadata reference

bnowok committed:
Update warning message for metadata version

bnowok committed:
Clarify parameter description in metadata function

Updated parameter documentation for clarity.

bnowok committed:
Fix roxygen2 configuration in DESCRIPTION file

bnowok committed:
Style code (GHA)

bnowok committed:
Document package (GHA)

bnowok committed:
Fix comment formatting in metadata.R

bnowok committed:
Style code (GHA)

bnowok committed:
Clean up comments in metadata function

bnowok committed:
Document package (GHA)

bnowok reviewed (approved):
I've made some changes to the wording only
